### PR TITLE
ArPow stage 1: local source-build infrastructure

### PR DESCRIPTION
**Pull request by MichaelSimons**
This PR adds the local build infrastructure that lets ArPow (arcade-powered source-build) run in this repo. See <https://github.com/dotnet/source-build/blob/master/Documentation/planning/arcade-powered-source-build/onboarding/local-onboarding.md> for more details about how it works.

To try it out locally, run this on Linux: `./build.sh -c Release --restore --build --pack /p:ArcadeBuildFromSource=true -bl`

This PR should have no effect on ordinary builds, or CI. ArPow stage 2 will add source-build to CI: PR validation and official builds.

For <https://github.com/dotnet/source-build/blob/master/Documentation/planning/arcade-powered-source-build/implementation-plan.md>

Related to https://github.com/dotnet/source-build/issues/2074

### CODE REVIEW

**Review by brettfo — APPROVED** [review]

File: eng/common/templates/steps/source-build.yml [hunk 1]
```diff
@@ -36,7 +36,7 @@ steps:
 
     ${{ coalesce(parameters.platform.buildScript, './build.sh') }} --ci \
       --configuration $buildConfig \
-      --restore --build --pack --publish \
+      --restore --build --pack --publish -bl\
```
Comment (MichaelSimons) [inline eng/common/templates/steps/source-build.yml:39]: This needs to get pushed into Arcade - https://github.com/dotnet/source-build/issues/2090

Comment (MichaelSimons) [reply eng/common/templates/steps/source-build.yml:39]: https://github.com/dotnet/arcade/pull/7104

**Comment by MichaelSimons** [comment]
@dseefeld, please review at your convenience.

File: eng/SourceBuild.props [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    <PropertyGroup>
+      <InnerBuildArgs>$(InnerBuildArgs) /p:Projects="$(InnerSourceBuildRepoRoot)\FSharp.sln"</InnerBuildArgs>
+    </PropertyGroup>
+  </Target>
+
+  <!--
+    The build script bootstraps some tooling for the build.  Since the inner build is triggerred via msbuild,
+    trigger the bootstrapping for the inner build.
+  -->
+  <Target Name="CopyBootstrap"
```
Comment (dseefeld) [inline eng/SourceBuild.props:37]: It's not really copying, it's building some bootstrap tooling, right?  Maybe name it "BuildBoostrap"?

**Review by dseefeld — APPROVED** [review]

File: eng/SourceBuild.props (hunk 2 above)
Comment (MichaelSimons) [reply eng/SourceBuild.props]: Good point,  this was a leftover from an earlier iteration where I tried coping the bootstrapped binaries into the inner build.  I renamed this target as you suggested.

File: eng/common/templates/steps/source-build.yml (hunk 1 above)
Comment (MichaelSimons) [reply eng/common/templates/steps/source-build.yml:39]: The arcade change is in place but has not been picked up in this repo yet.  Waiting for https://github.com/dotnet/fsharp/pull/11306 to be merged.

Comment (MichaelSimons) [reply eng/common/templates/steps/source-build.yml:39]: The latest version of Arcade was merged in, this dependency is now unblocked.

**Comment by MichaelSimons** [comment]
@brettfo, this PR is ready to go.  Could you please merge?

**Review by cartermp — APPROVED** [review]

**Comment by cartermp** [comment]
Thanks!

**Pull request merged**